code/modules/mob/living/silicon/robot/robot_modules.dm

```diff
@@ -231,6 +233,40 @@
 	if(R.hud_used)
 		R.hud_used.update_robot_modules_display()
 	SSblackbox.record_feedback("tally", "cyborg_modules", 1, R.module)
+	
+/obj/item/robot_module/proc/analyze_reagents(var/list/reagents_inside)
+	var/mob/living/silicon/robot/R = loc
+	to_chat(loc, "<span class='notice'>You start analyzing the solution.</span>")
+	var/i = 0
+	for(var/datum/reagent/each_reagent in reagents_inside) //If there are multiple reagents
+		i += 10 //So it kinda prevents spam
+		if(each_reagent in learned_reagents) //If you can already synthetize it, no need to analyze again
+			addtimer(CALLBACK(GLOBAL_PROC ,/proc/to_chat, loc, "<span class='notice'>You already successefully analyzed [each_reagent].</span>"), i)
+		else
+			addtimer(CALLBACK(GLOBAL_PROC ,/proc/to_chat, loc, "<span class='notice'>Analyzing unknown reagent.</span>"), i)
+			R.cell.use(15) //Consumes energy to analyze. Not sure what happens if the borg do not have enough energy. Will consume energy even if it fails to analyze
+			addtimer(CALLBACK(GLOBAL_PROC ,/proc/to_chat, loc, "<span class='notice'>You successefully analyzed [each_reagent]!.</span>"), 300 + i) //Add a timer to simulate analyzing
+			if(!each_reagent.can_synth)
+				addtimer(CALLBACK(GLOBAL_PROC ,/proc/to_chat, loc, "<span class='notice'>But it is too complex.</span>"), 300 + i) //It's not like it can know beforehand what it is
+			else if(each_reagent in whitelisted_reagents || R.emagged || whitelisted_reagents == null) //Only allowed reagents. Can be set by borg type. Except if emmaged. If empty then allow everything
+				addtimer(VARSET_CALLBACK(src, learned_reagents, learned_reagents.Add(list(each_reagent))), 300) //Add reagent to list of things that can be synthetized
+			else
+				addtimer(CALLBACK(GLOBAL_PROC ,/proc/to_chat, loc, "<span class='notice'>But it is forbidden, so you delete the data.</span>"), 300 + i) //Same as above
+		
+/obj/item/robot_module/proc/synthetize_reagents()
+	var/mob/living/silicon/robot/R = loc
+	if(learned_reagents == null) //You know nothing Jon Snow

```

Comment:
The !learned_reagents.len is what I need! Thanks!

About the `M==user || IsCyborg(user)` is because a borg cannot drink. So it is excluded from the if check(I tried drinking as a bro borg and couldn't). By adding the Or it goes on, but then I have to have separate interactions for humans and borg.  This way, you drink if you are a human targeting yourself, you drink if you are a human targeting a borg(borgs dont drink), you analyze if you are a borg and target yourself, or you analyze if you are a borg and target another borg. Its hacky, but I fear that if I mess with the attack thing I break more than I add :/. 

I am working on the timer proc suggestion @XDTM made

Reply:
If i'm understanding correctly your organization needs work

if `M == user` will never work because a borg cannot drink, then do not add the `Iscyborg` to the check just so you can ifelse below.

**if the above is true** (I think i'm just misunderstanding you), you should have

```DM
if(M == user) // Your normal check
    blah blah blah
if(iscyborg(user)) //Cyborg-specific interaction because above check fails (?)
    blah blah blah2
```

Reply:
OH. Yes! Got it! Will do it.